### PR TITLE
feat(launchpad): detect if runner is set up and change UI accordingly

### PR DESCRIPTION
**Pull request by lmiller1990**
https://cypress-io.atlassian.net/browse/UNIFY-292

Requirements of ticket:

- changes the UI to show "LAUNCH" if the testing type (CT or E2E) is configured
- otherwise it says "Click here to configure... "
- clicking "LAUNCH" launches the CT runner
  - At this point in time, a testing type is considered "configured" if you have at least one property inside of the `component` or `e2e` keys inside of `cypress.json`. This will change when #17000 is merged (where we will use the `devServer` or `setupNode` key sin `plugins` to figure out if CT/E2E has been initialized).

Some other things were added to get launchpad to be a bit more functional in general:

- added some code to cover some edge cases. Namely, sometimes the electron app can start before graphql has initialized the project (by initialize, I mean resolve the configuration and initialize plugins). We poll until the graphql server has started, and the current project has been initalized (config, plugins)
  - in the future, we probably won't want to initialize plugins until the user actually selects a runner. At this point, everything is very tightly coupled by the `open_project` singleton, and will take a bit of heavy lifting to uncouple. This will be okay until we have something a bit better.

Using launchpad as an example, since we have configured CT, it shows "launch" for the CT card. E2E has not been configured, so it shows "Click here to configure ...". These are not the final labels, just placeholders.

image

### CODE REVIEW

File: circle.yml [hunk 1]
```diff
@@ -386,7 +386,7 @@ commands:
         type: string
     steps:
       - restore_cached_workspace
-      - run: yarn workspace @packages/launchpad cypress:run --browser <<parameters.browser>>
+      - run: yarn workspace @packages/launchpad build && yarn workspace @packages/launchpad cypress:run --browser <<parameters.browser>>
```
Comment (tgriesser) [inline circle.yml:389]: Turns out `build` is not necessary here, I was wrong about the root cause of the flake.

File: packages/graphql/schema.graphql [hunk 2]
```diff
@@ -66,6 +66,15 @@ scalar JSON @specifiedBy(url: "http://www.ecma-international.org/publications/fi
...
 type LocalProject {
+  """
+  Whether the user has configured component testing. Based on the existance of a 'component' key in their cypress.json
+  """
+  hasSetupComponentTesting: Boolean!
+
+  """
+  Whether the user has configured e2e testing or not, based on the existance of a 'component' key in their cypress.json
+  """
+  hasSetupE2ETesting: Boolean!
```
Comment (tgriesser) [inline packages/graphql/schema.graphql:77]: Nice, love the descriptive graphql fields 👍

File: packages/launchpad/src/setup/TestingTypeCards.vue [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+fragment TestingTypeCardsApp on App {
+  activeProject {
+    hasSetupComponentTesting
+    hasSetupE2ETesting
+  }
+}
+`
+  
+gql`
+fragment TestingTypeCardsWizard on Wizard {
```
Comment (tgriesser) [inline packages/launchpad/src/setup/TestingTypeCards.vue:63]: Any reason these two couldn't be consolidated into:

```graphql
fragment TestingTypeCards on Query {
  app {
    activeProject {
      hasSetupComponentTesting
      hasSetupE2ETesting
    }
  }
  wizard {
    testingTypes {
      id
      title
      description
    }
  }
}
```

File: packages/launchpad/src/setup/TestingTypeCards.vue [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+
+const emit = defineEmits<{
+  (event: 'launchCt'): void
+  (event: 'launchE2E'): void
+}>()
+
+const ct = computed(() => {
+  const type = props.gql?.wizard?.testingTypes?.find(x => x.id === 'component')
+  if (!type) {
+    throw Error('Did not get testingType from API')
```
Comment (tgriesser) [inline packages/launchpad/src/setup/TestingTypeCards.vue:100]: Is this possible and if so, should we make the fields non-null so it's not?

File: packages/launchpad/src/App.vue [hunk 5]
```diff
@@ -1,18 +1,66 @@
...
+      window.clearInterval(interval)
+    } else {
+      query.executeQuery()
+    }
+  } catch (e) {
+    // probably graphql server has not initialized yet
+  }
+}
+
+interval = window.setInterval(poll, 200)
```
Comment (tgriesser) [inline packages/launchpad/src/App.vue:63]: Interesting, I wonder if there's a better way we should be handling things on the server so this race condition isn't a frontend concern.

**Review by tgriesser — COMMENTED** [review]
Left a few comments - also have been wondering if there's value to the `App` type, or if we should just move those fields to the top level `Query`

File: packages/launchpad/src/App.vue (hunk 5 above)
Comment (lmiller1990) [reply packages/launchpad/src/App.vue:63]: I'd say so. If some activities need to occur before the UI is rendered, I suppose something like

```js
async function start (openElectron) {
  // do stuff
  await something()

  openElectron()
}
```

I'd say once we untangle the mess that is `open_project`, this problem will natural disappear.

File: packages/launchpad/src/setup/TestingTypeCards.vue (hunk 3 above)
Comment (lmiller1990) [reply packages/launchpad/src/setup/TestingTypeCards.vue:63]: I could not get `mountFragment` to work with Fragments on the top level `Query`. I wonder if I'm missing something.

File: packages/launchpad/src/setup/TestingTypeCards.vue (hunk 4 above)
Comment (lmiller1990) [reply packages/launchpad/src/setup/TestingTypeCards.vue:100]: Good idea

Comment (lmiller1990) [reply packages/launchpad/src/setup/TestingTypeCards.vue:100]: 595e8d02b43b7e66ba36c7f34f4a54baab7c5c6d

**Comment by lmiller1990** [comment]
@tgriesser made most of the recommended changes and we green, so going to merge. We should explore:

1. https://github.com/cypress-io/cypress/pull/17934#discussion_r698688476 - `mountFragment` does not seem to type correctly with top-level `Query`. I tried fixing it but need to spend a lot more time to really get how the mount codegen works.
2. Better way to spin everything up and solve a potential race condition - [as you noted](https://github.com/cypress-io/cypress/pull/17934#discussion_r698877394) this isn't ideal.

**Pull request merged**